### ISSUE TITLE
Dynamically extending ImportExport column list

### ISSUE THREAD

**Issue by Flynsarmy** [octobercms/october#2029]
### The issue

I am using the ImportExportController behavior in my plugin however the columns to import and export must be dynamically generated at runtime based on the current backend URL. There doesn't appear to be a way to extend the list columns at runtime. See the current implementation [here](https://github.com/octobercms/october/blob/master/modules/backend/behaviors/ImportExportController.php#L667).

### Proposal

I'd like to add a *importExportExtendListColumnConfig* controller method to handle this. The method name is ugly and can be changed. Here is my proposed implementation:

    protected function makeListColumns($config)
    {
        $config = $this->controller->importExportExtendListColumnConfig($config);
        $config = $this->makeConfig($config);

        if (!isset($config->columns) || !is_array($config->columns)) {
            return null;
        }
        ...
    }

    /**
     * Called after the list columns are defined.
     * @param mixed $config The current import/export columns
     * @return mixed
     */
    public function importExportExtendListColumnConfig($config)
    {
        return $config;
    }

The `makeListColumns` method is used by both [getImportDbColumns](https://github.com/octobercms/october/blob/master/modules/backend/behaviors/ImportExportController.php#L281) and [getExportColumns](https://github.com/octobercms/october/blob/master/modules/backend/behaviors/ImportExportController.php#L455) so perhaps adding a `$context` argument to `makeListColumns` which is then passed to `importExportExtendListColumnConfig` would be helpful too.

### Example Usage

In a controller implementing ImportExportController behavior:

    public function importExportExtendListColumnConfig($config)
    {
        $config = Yaml::parse(File::get($config));
        $config['columns']['my_new-column'] = ['label' => 'My New Column'];

        return $config;
    }

If you're happy with this I can write up a pull request.

**Comment by daftspunk** [comment]
This sounds right, except maybe the method should probably be called `importExportExtendColumns`. We need to understand more of the *why* instead of the *how*. Perhaps there is a more elegant way to solve the problem, but without knowing the problem it is difficult to know.

**Comment by EdisonLV** [comment]
Not sure if this solves the problem completely, but I think it is starting point...
**Extend your controller**
~~~
    public function import()
    {

        if (Request::url() == 'example.oi') {
            $this->importColumns = [
                'column_name' => 'Column Attribute'
            ];            
        }


        // Call the ListController behavior index() method
        $this->asExtension('ImportExportController')->import();
    }
~~~

**Comment by jeremymouton** [comment]
This is an old topic, but I was able to extend a plugin's export columns by overwriting the config file entirely.

```php
\Acme\MyPlugin\Controllers\MyController::extend(function($controller) {
  $controller->importExportConfig = '$/acme2/mypluginextension/controllers/mycontroller/config_import_export.yaml';
});
```

**Comment by daftspunk** [comment]
Thanks @jeremymouton

There isn't still a clear use case for this, however, we've added some basic overrides in v2.1.11

    /**
     * importExportExtendExportColumns
     */
    public function importExportExtendExportColumns($columns)
    {
        return $columns;
    }

    /**
     * importExportExtendImportColumns
     */
    public function importExportExtendImportColumns($columns)
    {
        return $columns;
    }

**Comment by SebastiaanKloos** [comment]
> Thanks @jeremymouton
> 
> There isn't still a clear use case for this, however, we've added some basic overrides in v2.1.11
> 
> ```
> /**
>  * importExportExtendExportColumns
>  */
> public function importExportExtendExportColumns($columns)
> {
>     return $columns;
> }
> 
> /**
>  * importExportExtendImportColumns
>  */
> public function importExportExtendImportColumns($columns)
> {
>     return $columns;
> }
> ```

Hi @daftspunk, There is currently no way to use the function. You can't override the function in the controller consuming the behavior.

Maybe this can be implemented in the behavior.

``` php
$columns = $this->controller->importExportExtendImportColumns($columns);
```

What do you think?

**Comment by daftspunk** [comment]
Good catch @SebastiaanKloos - fixed in v2.1.12

Thank you!
